### DISCUSSION TITLE
Is there a 2d drawing library for avalonia, like Konva for JS, where more complex drawing applications can be built on?

### DISCUSSION THREAD

**Discussion by lanyusan** [AvaloniaUI/Avalonia#9614] [Q&A] [answered]
In Web/JS eco system, there are some really good 2d drawing libraries with builtin controls for rotation, resizing, grouping, animation etc. and also plenty of builtin 2d objects. I have used Konva Js, https://konvajs.org/, which is a really great library.

I am wondering if there is such a library out there for Avalonia? 

I have done some research and haven't found one yet.

**Comment by robloo** [comment, marked as answer]
Avalonia itself has powerful drawing and animation capabilities that should be used first. Take a look at the Shapes such as Ellipse, Rectangle, Arc, Path, etc. These can be drawn inside of a Canvas control.

If you want to go lower level, Avalonia is rendered using Skia. So you can also render inside Avalonia using SkiaSharp. This is fully cross platform at the moment as Skia is the renderer used everywhere. However, if the rendering library is ever changed in the future this might not be the best approach.

**Reply by lanyusan** [reply]
Yes I am aware of that. 

Drawing shapes is easy. It is just quite complicated to develop a whole suite of drawing controls, such zindex of overlapping shapes, grouping of shapes, rotation, resizing, drag and drop etc. A whole tool set needs to be developed.

**Reply by robloo** [reply]
I think there are no such high-level 2D graphics libraries like what you are looking for. Even in other XAML/C# frameworks I'm not aware of any library like this.

Several applications implement this sort of functionality themselves though. There is Lunacy by Icons8 and also Core2D by @wieslawsoltes. The respective authors might be more help.

**Reply by lanyusan** [reply]
@robloo  Thanks. I have actually looked at draw2d / core2d by @wieslawsoltes though haven't really gotten deeply into how he did it.

Would certain seek advice from the community if we eventually decide to migrate our legacy Electron app to avalonia.

**Reply by beto-rodriguez** [reply]
I didn't know there was a demand on this, I built a small framework, that uses SkiaSharp to render geometries and controls across multiple platforms, (SkiaSharp should not be required but *The code needs some fixes to run on another engine*), it is small, but it was necessary to build LiveCharts2, there are some issues I am aware of, but this is maybe a good starting point?

[This](https://github.com/beto-rodriguez/LiveCharts2/blob/dev/samples/ViewModelsSamples/General/TemplatedLegends/CustomLegend.cs#L68) is a sample of a small drawn control (the tooltip of the charts), that code can run across Avalonia or windows forms or wpf or maui or Uno or WinUI, I mean all the supported platforms of the library.

Geometries and controls support animations, transforms, easing functions, but there are no docs yet about the framework behind yet.

**Reply by timunie** [reply]
> Would certain seek advice from the community if we eventually decide to migrate our legacy Electron app to avalonia.

While you can ask the community and you often get support there for free, there is no guarantee that anyone can answer your question in time. If you need more reliable support, consider to use the paid support (Especially if it's a commercial App). This is just a hint, don't feel forced to do.

**Reply by lanyusan** [reply]
@beto-rodriguez Thanks for the advice. I would study your code definitely.

As for if there is a market for a library similar to Konva JS for Avalonia/Maui, I am pretty certain there is, though not sure how big it is.

If there is such a Avalonia/Maui library close to Konva JS, a lot of specialized design tools targeting certain industry could be developed if we don't need to develop those drawing control tool sets from ground up. 

I have personally tried to build such tool set with JS from ground up and it took tremendous effort to build something functioning, but still not even close to what Konva JS can offer.
